### PR TITLE
ReplicaSet controller can set/remove ControllerRef

### PR DESCRIPTION
**Pull request by wojtek-t**
This is mostly a copy from https://github.com/kubernetes/kubernetes/pull/27600

### CODE REVIEW

**Comment by caesarxuchao** [comment]
```
FAIL	k8s.io/kubernetes/contrib/mesos/cmd/km [build failed]
```
Compilation error in this directory is only exposed when running unit tests.

**Comment by caesarxuchao** [comment]
There is a integration test in #27600, if you want to copy that over as well. Note that the 27600 version has a flake, so please copy the latest file :)

**Comment by wojtek-t** [comment]
@caesarxuchao - sure - I can add integration test too.

Other than that, do you have any other comments?

**Comment by caesarxuchao** [comment]
No, otherwise LGTM. Feel free to apply the lgtm. Thanks!

**Comment by wojtek-t** [comment]
@caesarxuchao - I've sent out https://github.com/kubernetes/kubernetes/pull/29275 out for review. This will slightly unify RS and RC. And with that it would be much easier to actually copy the integration tests (integration tests are taking advantage of the shared pod informer).

Once that PR is merged, I will update this one.

**Comment by wojtek-t** [comment]
I've update the PR, but this is currently blocked on #29275 - it would be great if you could look on that one too.

File: pkg/controller/replicaset/replica_set.go [hunk 1]
```diff
@@ -542,16 +575,60 @@ func (rsc *ReplicaSetController) syncReplicaSet(key string) error {
...
+		podList, err := rsc.podStore.Pods(rs.Namespace).List(labels.Everything())
+		if err != nil {
+			glog.Errorf("Error getting pods for rs %q: %v", key, err)
+			rsc.queue.Add(key)
+			return err
+		}
+		cm := controller.NewPodControllerRefManager(rsc.podControl, rs.ObjectMeta, selector, getRSKind())
+		matchesAndControlled, matchesNeedsController, controlledDoesNotMatch := cm.Classify(podList.Items)
+		for _, pod := range matchesNeedsController {
+			err := cm.AdoptPod(pod)
```
Comment (kargakis) [inline pkg/controller/replicaset/replica_set.go]: If this is mutating you need to deep-copy the pod beforehand.

Comment (caesarxuchao) [inline pkg/controller/replicaset/replica_set.go]: Thanks. I took another look and it's not mutating pod.

Comment (kargakis) [inline pkg/controller/replicaset/replica_set.go]: Same thing for `ReleasePod`?

Comment (caesarxuchao) [inline pkg/controller/replicaset/replica_set.go]: Yes, both of them builds patch based on the pod and sends the patch to the server.

**Comment by wojtek-t** [comment]
@caesarxuchao - this is now ready for final review; PTAL

**Comment by caesarxuchao** [comment]
LGTM. Thank you @wojtek-t!

**Pull request merged**